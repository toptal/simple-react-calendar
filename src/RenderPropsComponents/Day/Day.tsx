import React, { FC } from 'react'
import cx from 'classnames'

import * as helper from '../../helper'
import { HandleOnDayClick, HandleOnDayEnter } from '../../@types'

interface Props {
  blockClassName: string
  date: Date
  getDayFormatted: typeof helper.getDayFormatted
  handleOnClick: HandleOnDayClick
  handleOnEnter: HandleOnDayEnter
  isCurrentMonth: boolean
  isDisabled: boolean
  isHighlighted: boolean
  isMonthNext: boolean
  isMonthPrev: boolean
  isNonSelectable: boolean
  ISODate: string
  isSelectable: boolean
  isSelected: boolean
  isSelectionEnd: boolean
  isSelectionStart: boolean
  isToday: boolean
  isWeekend: boolean
  isWorkDay: boolean
  key?: string
}

const Day: FC<Props> = ({
  blockClassName,
  date,
  getDayFormatted,
  handleOnClick,
  handleOnEnter,
  isCurrentMonth,
  isDisabled,
  isHighlighted,
  isMonthNext,
  isMonthPrev,
  isNonSelectable,
  ISODate,
  isSelectable,
  isSelected,
  isSelectionEnd,
  isSelectionStart,
  isToday,
  isWeekend,
  isWorkDay
}) => (
  <button
    data-simple-react-calendar-day={ISODate}
    className={cx(`${blockClassName}-week-day`, {
      'is-current_month': isCurrentMonth,
      'is-disabled': isDisabled,
      'is-end_selection': isSelectionEnd,
      'is-highlighted': isHighlighted,
      'is-next_month': isMonthNext,
      'is-not_selectable': isNonSelectable,
      'is-prev_month': isMonthPrev,
      'is-selectable': isSelectable,
      'is-selected': isSelected,
      'is-start_selection': isSelectionStart,
      'is-today': isToday,
      'is-weekend': isWeekend,
      'is-working_day': isWorkDay
    })}
    onClick={handleOnClick}
    onMouseEnter={handleOnEnter}
    type='button'
    value={ISODate}
  >
    {getDayFormatted(date)}
  </button>
)

Day.displayName = 'Day'

export default Day
