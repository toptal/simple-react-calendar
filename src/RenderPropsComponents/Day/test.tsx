import { fireEvent, render } from '@testing-library/react'
import React, { ComponentProps } from 'react'

import { getDayFormatted } from '../../helper'
import Day from '.'

const renderDay = (props: Partial<ComponentProps<typeof Day>>) => {
  const baseProps = {
    ISODate: '2020-02-09',
    blockClassName: 'example-class-name',
    date: new Date(2020, 1, 9),
    getDayFormatted,
    handleOnClick: jest.fn(),
    handleOnEnter: jest.fn(),
    isCurrentMonth: true,
    isDisabled: false,
    isHighlighted: true,
    isMonthNext: false,
    isMonthPrev: false,
    isNonSelectable: true,
    isSelectable: true,
    isSelected: false,
    isSelectionEnd: false,
    isSelectionStart: true,
    isToday: true,
    isWeekend: true,
    isWorkDay: false,
    ...props
  }

  // eslint-disable-next-line react/jsx-props-no-spreading
  return render(<Day {...baseProps} />)
}

test('default render', () => {
  const { container } = renderDay({})

  expect(container.firstChild).toMatchSnapshot()
})

test('trigger callback on click', () => {
  const handleOnClick = jest.fn()
  const { getByRole } = renderDay({ handleOnClick })

  const button = getByRole('button')

  fireEvent.click(button)

  expect(handleOnClick).toBeCalled()
})

test('trigger callback on mouseenter', () => {
  const handleOnEnter = jest.fn()
  const { getByRole } = renderDay({ handleOnEnter })

  const button = getByRole('button')

  fireEvent.mouseEnter(button)

  expect(handleOnEnter).toBeCalled()
})

test('handle currentMonth', () => {
  const { container } = renderDay({ isCurrentMonth: true })

  expect(container.firstChild).toHaveClass('is-current_month')
})

test('handle isDisabled', () => {
  const { container } = renderDay({ isDisabled: true })

  expect(container.firstChild).toHaveClass('is-disabled')
})

test('handle isHighlighted', () => {
  const { container } = renderDay({ isHighlighted: true })

  expect(container.firstChild).toHaveClass('is-highlighted')
})

test('handle isMonthNext', () => {
  const { container } = renderDay({ isMonthNext: true })

  expect(container.firstChild).toHaveClass('is-next_month')
})

test('handle isMonthPrev', () => {
  const { container } = renderDay({ isMonthPrev: true })

  expect(container.firstChild).toHaveClass('is-prev_month')
})

test('handle isNonSelectable', () => {
  const { container } = renderDay({ isNonSelectable: true })

  expect(container.firstChild).toHaveClass('is-not_selectable')
})

test('handle isSelectable', () => {
  const { container } = renderDay({ isSelectable: true })

  expect(container.firstChild).toHaveClass('is-selectable')
})

test('handle isSelected', () => {
  const { container } = renderDay({ isSelected: true })

  expect(container.firstChild).toHaveClass('is-selected')
})

test('handle isSelectionEnd', () => {
  const { container } = renderDay({ isSelectionEnd: true })

  expect(container.firstChild).toHaveClass('is-end_selection')
})

test('handle isSelectionStart', () => {
  const { container } = renderDay({ isSelectionStart: true })

  expect(container.firstChild).toHaveClass('is-start_selection')
})

test('handle isToday', () => {
  const { container } = renderDay({ isToday: true })

  expect(container.firstChild).toHaveClass('is-today')
})

test('handle isWeekend', () => {
  const { container } = renderDay({ isWeekend: true })

  expect(container.firstChild).toHaveClass('is-weekend')
})

test('handle isWorkday', () => {
  const { container } = renderDay({ isWorkDay: true })

  expect(container.firstChild).toHaveClass('is-working_day')
})
