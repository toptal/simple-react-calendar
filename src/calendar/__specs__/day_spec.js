import React from 'react'

import ReactTestRenderer from 'react-test-renderer'
import Day from '../day'

import {shallow} from 'enzyme'

describe('Day', () => {
  let defaultProps, event, tree, props, wrapper

  beforeEach(() => {
    event = {preventDefault: jest.fn()}
    defaultProps = {
      date: new Date(2015, 7, 17),
      today: new Date(2015, 7, 17),
      activeMonth: new Date(2015, 7, 17),
      onClick: () => {},
      onMouseMove: () => {},
    }
  })

  describe('#_onClick', () => {
    context('when prop `onClick` is defined', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {onClick: jest.fn()})
        wrapper = shallow(<Day {...props} />)

        wrapper.instance()._onClick(event)

        expect(event.preventDefault).toHaveBeenCalledTimes(1)
        expect(props.onClick).toHaveBeenCalledWith(props.date)
      })
    })

    context('when prop `onClick` is undefined', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {onClick: undefined})
        wrapper = shallow(<Day {...props} />)

        wrapper.instance()._onClick(event)

        expect(event.preventDefault).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('#_onMouseMove', () => {
    context('when prop `onMouseMove` is defined', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {onMouseMove: jest.fn()})
        wrapper = shallow(<Day {...props} />)

        wrapper.instance()._onMouseMove(event)

        expect(event.preventDefault).toHaveBeenCalledTimes(1)
        expect(props.onMouseMove).toHaveBeenCalledWith(props.date)
      })
    })

    context('when prop `onMouseMove` is undefined', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {onMouseMove: undefined})
        wrapper = shallow(<Day {...props} />)

        wrapper.instance()._onMouseMove(event)

        expect(event.preventDefault).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('#render', () => {
    it('renders <Day />', () => {
      tree = ReactTestRenderer.create(<Day {...defaultProps} />).toJSON()

      expect(tree).toMatchSnapshot()
    })

    context('when prop `today` is different from prop `date`', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {date: new Date(2015, 8, 2)})
        tree = ReactTestRenderer.create(<Day {...props} />).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    context('when prop `blockClassName` is defined', () => {
      it('renders <Day />', () => {
        props = Object.assign(defaultProps, {blockClassName: 'cal'})
        tree = ReactTestRenderer.create(<Day {...props} />).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })
  })
})
