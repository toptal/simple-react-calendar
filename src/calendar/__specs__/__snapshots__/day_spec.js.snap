// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day #render renders <Day /> 1`] = `
<div
  className="calendar-day"
  onClick={[Function]}
  onMouseMove={[Function]}
>
  17
</div>
`

exports[`Day #render when prop \`blockClassName\` is defined renders <Day /> 1`] = `
<div
  className="cal-day"
  onClick={[Function]}
  onMouseMove={[Function]}
>
  17
</div>
`

exports[`Day #render when prop \`today\` is different from prop \`date\` renders <Day /> 1`] = `
<div
  className="calendar-day"
  onClick={[Function]}
  onMouseMove={[Function]}
>
  2
</div>
`
